src/shogun/base/SGObject.h

```diff
@@ -278,6 +282,131 @@ class CSGObject
+	 * @return true if the parameter exists with the input name
+	 */
+	inline bool has(const std::string& name) const
+	{
+		BaseTag tag(name);
+		return _has(tag);
+	}
+
+	/** Checks if a parameter with a tag already exists.
+	 *
+	 * @param tag tag of the parameter containing name and type information
+	 * @return true if the parameter exists with the input tag
+	 */
+	template <typename T>
+	inline bool has(const Tag<T>& tag) const
+	{
+		return has<T>(tag.name());
+	}
+
+	/** Checks if a type exists for a parameter identified by a name.
+	 *
+	 * @param name name of the parameter
+	 * @param type type information for the parameter
+	 * @return true if the parameter exists with the input name and type
+	 */
+	template <typename T, typename U=void>
+	inline bool has(const std::string& name) const
+	{
+		BaseTag tag(name);
+		if(!_has(tag))
+			return false;
+		const Any value = _get(tag);
+		return value.sameType<T>();
+	}
+
+	/** Sets a value for an allowed parameter which is identified by a tag.
+	 *
+	 * @param _tag name and type information of parameter
+	 * @param value value of the parameter
+	 */
+	template <typename T>
+	inline void set(Tag<T>& _tag, const T& value)
+	{
+		// check for name in tag
+		if(_has(_tag))
+		{
+			// check for type in tag
+			if(has<T>(_tag.name()))
+				_set(_tag, erase_type(value));
+			else
+			{
+				SG_SERROR("Type for parameter with name \"%s\" is not correct.\n",
+					_tag.name().c_str());
+			}
+		}
+		else
+		{
+			SG_SERROR("Parameter with name \"%s\" is not allowed in \"%s\".\n",
+				_tag.name().c_str(), get_name());
+		}
+	}
+
+	/** Sets a value for an allowed parameter which is identified by a name.
+	 *
+	 * @param name name of the parameter
+	 * @param value value of the parameter along with type information
+	 */
+	template <typename T, typename U=void>
+	inline void set(const std::string& name, const T& value)
+	{
+		Tag<T> tag(name);
+		set(tag, value);
+	}
+
+	/** Gets the value of a parameter corresponding to a tag.
+	 *
+	 * @param _tag name and type information of parameter
+	 * @return value of the parameter corresponding to the input tag
+	 */
+	template <typename T>
+	inline T get(Tag<T>& _tag) const
+	{
+		const Any value = _get(_tag);
+		try
+		{
+			return recall_type<T>(value);
+		}
+		catch(std::logic_error)
+		{
+			SG_SERROR("Parameter with the input name exists but the type is not correct.\n");
+		}
+	}
+
+	/** Gets the value of a parameter corresponding to a name.
+	 *
+	 * @param name name of the parameter
+	 * @return value of the parameter corresponding to the input name and type
+	 */
+	template <typename T, typename U=void>
+	inline T get(const std::string& name) const
+	{
+		BaseTag tag(name);
+		const Any value = _get(tag);
+		try
+		{
+			return recall_type<T>(value);
+		}
+		catch(std::logic_error)
+		{
+			SG_SERROR("Parameter with the input name exists but the type is not correct.\n");
+		}
+	}
+
+	/** Gets a list of names of all parameters stored in this object.
```

Comment:
Write doc for users, not developer

"Returns a list of all variables available for this class. These can be modified using [reference to get/set)"